pkg/referrer/referrer.go

```diff
@@ -0,0 +1,133 @@
+)
+
+const metadataNameInLayer = "image/image.boot"
+
+type referrer struct {
+	remote *remote.Remote
+}
+
+func newReferrer(keyChain *auth.PassKeyChain, insecure bool) *referrer {
+	return &referrer{
+		remote: remote.New(keyChain, insecure),
+	}
+}
+
+// checkReferrer fetches the referrers and parse out the nydus
+// image by specified manifest digest.
+// it's using distribution list referrers API.
+func (r *referrer) checkReferrer(ctx context.Context, ref string, manifestDigest digest.Digest) (*ocispec.Descriptor, error) {
+	handle := func() (*ocispec.Descriptor, error) {
+		// Create an new resolver to request.
+		fetcher, err := r.remote.Fetcher(ctx, ref)
+		if err != nil {
+			return nil, errors.Wrap(err, "get fetcher")
+		}
+
+		// Fetch image referrers from remote registry.
+		rc, _, err := fetcher.(remotes.ReferrersFetcher).FetchReferrers(ctx, manifestDigest)
+		if err != nil {
+			return nil, errors.Wrap(err, "fetch referrers")
+		}
+		defer rc.Close()
+
+		// Parse image manifest list from referrers.
+		var index ocispec.Index
+		bytes, err := io.ReadAll(rc)
+		if err != nil {
+			return nil, errors.Wrap(err, "read referrers")
+		}
+		if err := json.Unmarshal(bytes, &index); err != nil {
+			return nil, errors.Wrap(err, "unmarshal referrers index")
+		}
+		if len(index.Manifests) == 0 {
+			return nil, fmt.Errorf("empty referrer list")
+		}
+
+		// Prefer to fetch the last manifest and check
+		// if it is a nydus image.
+		rc, err = fetcher.Fetch(ctx, index.Manifests[0])
+		if err != nil {
+			return nil, errors.Wrap(err, "fetch referrers")
+		}
+		defer rc.Close()
+
+		var manifest ocispec.Manifest
+		bytes, err = io.ReadAll(rc)
+		if err != nil {
+			return nil, errors.Wrap(err, "read manifest")
+		}
+		if err := json.Unmarshal(bytes, &manifest); err != nil {
+			return nil, errors.Wrap(err, "unmarshal manifest")
+		}
+		if len(manifest.Layers) < 1 {
+			return nil, fmt.Errorf("invalid manifest")
+		}
+		metaLayer := manifest.Layers[len(manifest.Layers)-1]
+		if !label.IsNydusMetaLayer(metaLayer.Annotations) {
+			return nil, fmt.Errorf("invalid nydus manifest")
+		}
+
+		return &metaLayer, nil
+	}
+
+	desc, err := handle()
+	if err != nil && r.remote.RetryWithPlainHTTP(ref, err) {
+		return handle()
+	}
+
+	return desc, err
+}
+
+// fetchMetadata fetches and unpacks nydus metadata file to specified path.
+func (r *referrer) fetchMetadata(ctx context.Context, ref string, desc ocispec.Descriptor, metadataPath string) error {
+	handle := func() error {
+		// Create an new resolver to request.
+		resolver := r.remote.Resolve(ctx, ref)
+		fetcher, err := resolver.Fetcher(ctx, ref)
+		if err != nil {
+			return errors.Wrap(err, "get fetcher")
+		}
+
+		// Unpack nydus metadata file to specified path.
+		rc, err := fetcher.Fetch(ctx, desc)
+		if err != nil {
+			return errors.Wrap(err, "fetch nydus metadata")
+		}
+		defer rc.Close()
+
+		if err := remote.UnpackFromTar(rc, metadataNameInLayer, metadataPath); err != nil {
```

Comment:
Can we rename `UpanckFromTar` to `Unpack`? It also decompess the fetched content

Reply:
Changed `UpanckFromTar` to `Unpack`.
The downloaded artifact is put into snapshot directory, I think it will be GCed by `Cleanup` interface.